### PR TITLE
Updates to License to include both MIT and LPGL-3.0-or-later and update CUDA

### LINKED ISSUE #13: Dropping `conda_build_config.yaml`
It looks like there is a custom `conda_build_config.yaml`

https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/blob/e4011600566bf78df010dbce85315f9b62bdeb51/recipe/conda_build_config.yaml#L1-L10

Mentioning this as it [confused one of the migrators recently]( https://github.com/regro/cf-scripts/actions/runs/6963632203/job/18949562094#step:9:5517 ). This could potentially impact other migrators and version updates from the bot

Am wondering if there is another approach we can apply here to accomplish the same goal. Could you please share a bit more about what problem the custom `conda_build_config.yaml` is solving? Maybe we can come with a better approach together

### PR DESCRIPTION
**Pull request by lvotapka**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Fixes #13
Addresses https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/pull/12#discussion_r1402878952

### CODE REVIEW

**Comment by lvotapka** [comment]
I've made the changes to the meta.yaml file that should have your suggestions for including both the MIT and LPGL licenses, @jakirkham.

**Comment by lvotapka** [comment]
Alright here we go @jakirkham, this is the same error I was talking about. I don't even need to download the artifact to see that the problem is that Python cannot import the package - it cannot even import the packages on the CI itself. Here's where I got stuck last time - I cannot find in the logs why Python cannot import the package if compiled with CUDA 11.

**Comment by lvotapka** [comment]
For instance, I don't know if you can view it, but here's a link to one of the logs: https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/833921/logs/148

You can see on following lines, the unit tests fail because the package cannot be imported into Python:

```
2023-11-30T17:49:00.4856063Z Traceback (most recent call last):
2023-11-30T17:49:00.4857694Z   File "/home/conda/feedstock_root/build_artifacts/seekr2_openmm_plugin_1701366120978/test_tmp/run_test.py", line 5, in <module>
2023-11-30T17:49:00.4942254Z     import seekr2plugin
2023-11-30T17:49:00.4943544Z ModuleNotFoundError: No module named 'seekr2plugin'
```

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

That's even better

**Comment by jakirkham** [comment]
So here's what I'm seeing [on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=833921&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=431 ):

```bash
copying build/lib.linux-x86_64-3.10/seekr2plugin.py -> $BUILD_PREFIX/lib/python3.10/site-packages
copying build/lib.linux-x86_64-3.10/_seekr2plugin.cpython-310-x86_64-linux-gnu.so -> $BUILD_PREFIX/lib/python3.10/site-packages
```

This is interesting because `$BUILD_PREFIX` is where command line tools are. However `$PREFIX` is where package contents should be installed. So for some reason they are not getting installed where we would expect

**Comment by lvotapka** [comment]
First off, thanks @jakirkham for taking the time to look at this problem and help me figure out how to solve it.

Honestly, I'm quite confused about what exactly is the purpose of each of these sections (build, host, and run) mean in meta.yaml. I think that I just added openmm to all of them, because I always wanted OpenMM to be available for every stage of the conda-forge process. I will say that OpenMM has libraries that the seekr2_openmm_plugin needs in order to build. 

I'm willing to test any changes you feel might be helpful for resolving this problem. Another idea might be to remove the OpenMM 7.7 requirement - it's possible that version of OpenMM doesn't like CUDA 11...

**Comment by jakirkham** [comment]
Happy to help 🙂

Gotcha. The usual rule of thumb that I use for `requirements` is

* `build` is for build tools (so compilers, CMake, etc.) anything that needs to be executed during the build, but we don't want to link to
* `host` is for anything the build consumes (headers, libraries, etc.)
* `run` is for anything that should show up when a user installs the package (so Python or any libraries that were linked to)

Hope that is helpful. Please feel free to ask more questions

<hr>

Ok based on that here would be my suggestions

1. Let's drop `openmm` from `build`
2. Relax the version constraint on `openmm` (agree this is a good idea)

Think we can start with 1 and see how that goes. Then follow up on 2

It's possible other things may become clear as we iterate

**Comment by lvotapka** [comment]
Wow! Doing step 1 let all the checks pass - let me try again by removing the 7.7 dependency, and then downloading and testing one of the artifacts.

**Comment by lvotapka** [comment]
Very good news, @jakirkham, it seems that the builds are working, however I still cannot see any artifacts to download. Should we just proceed with merging this pull request, since the problem we had before seems to be gone?

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

Let's try re-rendering for good measure

@conda-forge-admin , please re-render

**Comment by lvotapka** [comment]
Let me try...

@conda-forge-admin , please re-render

**Comment by jakirkham** [comment]
Ah this is an org fork. Is that right?

Unfortunately there's not a good way to fix this (GitHub limitation)

Usual way to solve this is to [re-render locally]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-with-conda-smithy-locally )

Sorry about that 😞

**Comment by lvotapka** [comment]
No problem, it was easy to re-render locally.

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

Glad to hear that

<hr>

Noticed on CI there is a conflict regarding `numpy`

When looking at the recipe there are `requirements` lines in `host` and `run` that have `numpy` commented:

https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/blob/e4011600566bf78df010dbce85315f9b62bdeb51/recipe/meta.yaml#L53

https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/blob/e4011600566bf78df010dbce85315f9b62bdeb51/recipe/meta.yaml#L65

Though there is a line in `requirements/build` that has `numpy`:

https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/blob/e4011600566bf78df010dbce85315f9b62bdeb51/recipe/meta.yaml#L33

Could you please share some more context on this? Is `numpy` needed for this package? Or are these leftover from when `numpy` use to be a dependency? Was there some past issue when adding `numpy` as a dependency?

**Comment by lvotapka** [comment]
I do not believe that numpy should be a dependency, unless possibly when SWIG uses arrays. I think I was just including it because when I build OpenMM from source, it does require numpy.

Let me remove this dependency and try again...

File: recipe/meta.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
+  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win))]
```
Comment (jakirkham) [inline recipe/meta.yaml:19]: IIUC this is a CUDA only package. Is that right? If so, think we can make some simplifications

Also curious about the Windows Python 3.10 skip. It looks like that was related to the same CUDA issues ( https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/commit/6114afb8b54816bc1b23a76b110132952245d34f ). Is that right? If so, think we can try relaxing that

Making these changes would leave us with something like this

```suggestion
  skip: true  # [(cuda_compiler or "None") != "None"]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -30,20 +30,18 @@ requirements:
     - python                                 # [build_platform != target_platform]
     - cross-python_{{ target_platform }}     # [build_platform != target_platform]
     - cython                                 # [build_platform != target_platform]
-    - numpy                                  # [build_platform != target_platform]
     - swig                                   # [build_platform != target_platform]
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
     - {{ compiler('cuda') }}  # [linux64 or ppc641le or win]
```
Comment (jakirkham) [inline recipe/meta.yaml:36]: Think we can simplify this as well

```suggestion
    - {{ compiler('cuda') }}  # [(cuda_compiler or "None") != "None"]
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -16,7 +16,7 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
+  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win))]
   script_env:
     - CONFIG
     - CUDA_HOME  # [ppc64le]
```
Comment (jakirkham) [inline recipe/meta.yaml:22]: This should be preconfigured for CUDA 11. So think we can drop this

```suggestion
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -16,7 +16,7 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
+  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win))]
   script_env:
     - CONFIG
```
Comment (jakirkham) [inline recipe/meta.yaml:21]: What is this used for?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -61,7 +59,7 @@ requirements:
   run:
     - python
     - fftw
-    - openmm 7.7
+    - openmm
     #- {{ pin_compatible('numpy') }}
```
Comment (jakirkham) [inline recipe/meta.yaml:63]: Think we can drop this as `numpy` doesn't appear to be needed. Would also suggest dropping a similar line above (unfortunately GH won't let me comment on that one atm)

```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Lane! 🙏

Noticed some new CI failures. Think this has to do with how the `skip` is written. Made some suggestions on how to simplify that

Also included a couple other suggestions below

Would also recommend re-rendering after these changes are made (assuming they look reasonable)

Please let me know if you have any questions

File: recipe/meta.yaml (hunk 1 above)
Comment (lvotapka) [reply recipe/meta.yaml:19]: Yes, it is a CUDA only package, and I believe that indeed the Python 3.10 Windows skip was included for that reason. I'll make these changes and push in a moment...

File: recipe/meta.yaml (hunk 4 above)
Comment (lvotapka) [reply recipe/meta.yaml:21]: I am not sure, I had been using OpenMM's feedstock settings as a starting point, and I think I just included this because OpenMM included it as well.

Comment (jakirkham) [reply recipe/meta.yaml:21]: Ok I think [OpenMM is using this for some sort of testing check]( https://github.com/conda-forge/openmm-feedstock/commit/7f612876acd4576fd76698b1afe387871a1dc8dd ) that we don't have here. Let's go ahead and drop this

File: recipe/meta.yaml [hunk 6]
```diff
@@ -16,10 +16,9 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
+  skip: true  # [(cuda_compiler or "None") != "None"]
   script_env:
     - CONFIG
```
Comment (jakirkham) [inline recipe/meta.yaml:21]:
```suggestion
```

File: recipe/meta.yaml [hunk 7]
```diff
@@ -16,10 +16,9 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
+  skip: true  # [(cuda_compiler or "None") != "None"]
```
Comment (jakirkham) [inline recipe/meta.yaml:19]: Oops 🤦‍♂️ Sorry this should be `==`s

Please also re-render after this change

Should fix the latest CI errors

```suggestion
  skip: true  # [(cuda_compiler or "None") == "None"]
```

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

This looks pretty good. Thanks again for iterating on this with me 🙂

If you are happy with the changes here and CI passes, think we can bump the `build/number` and merge

**Comment by jakirkham** [comment]
Added a couple footnotes in the PR description and added a bit to the PR title

Hope that is ok

Also please feel free to change them as you see fit

**Comment by jakirkham** [comment]
Also one last thing, here are [the artifacts for the latest CI build]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=836531&view=artifacts&pathAsName=false&type=publishedArtifacts ). Please feel free to take a look if you would like to confirm things look ok. Happy to discuss if anything comes up

**Comment by jakirkham** [comment]
Was just noticing there is not currently a `bld.bat` for the Windows build. Do we want to add that or skip Windows for now? It looks like only Linux packages have been built for to-date. So maybe skipping Windows is reasonable. Mentioning in part as Windows CI fails for this reason

File: recipe/meta.yaml [hunk 8]
```diff
@@ -16,10 +16,7 @@ source:
 
 build:
   number: {{ build }}
-  skip: true  # [(win and py==310) or (cuda_compiler_version in (undefined, 'None') and (linux64 or win)) or (cuda_compiler_version != '10.2')]
-  script_env:
-    - CONFIG
-    - CUDA_HOME  # [ppc64le]
+  skip: true  # [(cuda_compiler or "None") == "None"]
```
Comment (jakirkham) [inline recipe/meta.yaml:19]: Suggesting only building for `linux-64` (consistent with what has been done so far)

```suggestion
  skip: true  # [not linux64]
  skip: true  # [(cuda_compiler or "None") == "None"]
```

Comment (lvotapka) [reply recipe/meta.yaml:19]: Yes, I'm also in favor of this.

**Comment by lvotapka** [comment]
Sounds good, @jakirkham, I will probably skip the Windows builds for now, and everything else except for linux64. I will download these artifacts tomorrow, and if everything looks good, I'll go ahead and merge.

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

Makes sense. Was realizing this was turning into unintentional scope creep. So wanted to dial back on that

Happy to revisit the other OSes/architectures in a separate issue/PR after (when you are ready). Though no need to if you would prefer not to

This already seems like a significant body of work for one PR. Thanks again for working on it with me 🙂

**Comment by jakirkham** [comment]
Thanks Lane! 🙏

It can take some time after packages are built for them to mirror to the CDN that conda-forge uses, which in turn can delay getting the latest package as a user

FWIW I did just try creating a fresh environment and it seemed to work ok

<details>

```
$ conda create --dry-run -n tst seekr2_openmm_plugin
Channels:
 - conda-forge
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
    current version: 23.10.0
    latest version: 23.11.0

Please update conda by running

    $ conda update -n base -c conda-forge conda



## Package Plan ##

  environment location: /opt/conda/envs/tst

  added / updated specs:
    - seekr2_openmm_plugin


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |            2_gnu          23 KB  conda-forge
    cudatoolkit-11.8.0         |      h4ba93d1_12       682.8 MB  conda-forge
    fftw-3.3.10                |nompi_hc118613_108         1.9 MB  conda-forge
    libblas-3.9.0              |20_linux64_openblas          14 KB  conda-forge
    libcblas-3.9.0             |20_linux64_openblas          14 KB  conda-forge
    libgfortran-ng-13.2.0      |       h69a702a_3          23 KB  conda-forge
    libgfortran5-13.2.0        |       ha4646dd_3         1.4 MB  conda-forge
    liblapack-3.9.0            |20_linux64_openblas          14 KB  conda-forge
    libopenblas-0.3.25         |pthreads_h413a1c8_0         5.3 MB  conda-forge
    numpy-1.26.2               |  py311h64a7726_0         7.7 MB  conda-forge
    ocl-icd-2.3.1              |       h7f98852_0         119 KB  conda-forge
    ocl-icd-system-1.0.0       |                1           4 KB  conda-forge
    openmm-8.1.0               |  py311h6d2dbb8_1        11.3 MB  conda-forge
    pcre-8.45                  |       h9c3ff4c_0         253 KB  conda-forge
    pthread-stubs-0.4          |    h36c2ea0_1001           5 KB  conda-forge
    python-3.11.6              |hab00c5b_0_cpython        29.3 MB  conda-forge
    python_abi-3.11            |          4_cp311           6 KB  conda-forge
    scipy-1.11.4               |  py311h64a7726_0        15.3 MB  conda-forge
    seekr2_openmm_plugin-0.1.6 |  py311h6818ba8_2         169 KB  conda-forge
    swig-4.0.2                 |       hd3c618e_2         1.2 MB  conda-forge
    ------------------------------------------------------------
                                           Total:       756.9 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge 
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu 
  bzip2              conda-forge/linux-64::bzip2-1.0.8-hd590300_5 
  ca-certificates    conda-forge/linux-64::ca-certificates-2023.11.17-hbcca054_0 
  cudatoolkit        conda-forge/linux-64::cudatoolkit-11.8.0-h4ba93d1_12 
  fftw               conda-forge/linux-64::fftw-3.3.10-nompi_hc118613_108 
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.40-h41732ed_0 
  libblas            conda-forge/linux-64::libblas-3.9.0-20_linux64_openblas 
  libcblas           conda-forge/linux-64::libcblas-3.9.0-20_linux64_openblas 
  libexpat           conda-forge/linux-64::libexpat-2.5.0-hcb278e6_1 
  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5 
  libgcc-ng          conda-forge/linux-64::libgcc-ng-13.2.0-h807b86a_3 
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-13.2.0-h69a702a_3 
  libgfortran5       conda-forge/linux-64::libgfortran5-13.2.0-ha4646dd_3 
  libgomp            conda-forge/linux-64::libgomp-13.2.0-h807b86a_3 
  liblapack          conda-forge/linux-64::liblapack-3.9.0-20_linux64_openblas 
  libnsl             conda-forge/linux-64::libnsl-2.0.1-hd590300_0 
  libopenblas        conda-forge/linux-64::libopenblas-0.3.25-pthreads_h413a1c8_0 
  libsqlite          conda-forge/linux-64::libsqlite-3.44.2-h2797004_0 
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-13.2.0-h7e041cc_3 
  libuuid            conda-forge/linux-64::libuuid-2.38.1-h0b41bf4_0 
  libzlib            conda-forge/linux-64::libzlib-1.2.13-hd590300_5 
  ncurses            conda-forge/linux-64::ncurses-6.4-h59595ed_2 
  numpy              conda-forge/linux-64::numpy-1.26.2-py311h64a7726_0 
  ocl-icd            conda-forge/linux-64::ocl-icd-2.3.1-h7f98852_0 
  ocl-icd-system     conda-forge/linux-64::ocl-icd-system-1.0.0-1 
  openmm             conda-forge/linux-64::openmm-8.1.0-py311h6d2dbb8_1 
  openssl            conda-forge/linux-64::openssl-3.2.0-hd590300_1 
  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0 
  pip                conda-forge/noarch::pip-23.3.1-pyhd8ed1ab_0 
  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h36c2ea0_1001 
  python             conda-forge/linux-64::python-3.11.6-hab00c5b_0_cpython 
  python_abi         conda-forge/linux-64::python_abi-3.11-4_cp311 
  readline           conda-forge/linux-64::readline-8.2-h8228510_1 
  scipy              conda-forge/linux-64::scipy-1.11.4-py311h64a7726_0 
  seekr2_openmm_plu~ conda-forge/linux-64::seekr2_openmm_plugin-0.1.6-py311h6818ba8_2 
  setuptools         conda-forge/noarch::setuptools-68.2.2-pyhd8ed1ab_0 
  swig               conda-forge/linux-64::swig-4.0.2-hd3c618e_2 
  tk                 conda-forge/linux-64::tk-8.6.13-noxft_h4845f30_101 
  tzdata             conda-forge/noarch::tzdata-2023c-h71feb2d_0 
  wheel              conda-forge/noarch::wheel-0.42.0-pyhd8ed1ab_0 
  xz                 conda-forge/linux-64::xz-5.2.6-h166bdaf_0 



DryRunExit: Dry run. Exiting.
```

</details>

If you are still seeing this error, could you please fill out [this issue template]( https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/issues/new?assignees=&labels=bug&projects=&template=1-bug-report.yml ) with info were the error was seen? That way we can dig into what is going on a bit more

**Comment by lvotapka** [comment]
Hey you're right! I just confirmed that it works - although for some reason it wants to pull the wrong version of openmm (version 8.0) when it installs, but that's OK, I can specify the users to install OpenMM 8.1, manually:

```
conda install -c conda-forge seekr2_openmm_plugin openmm=8.1
```

THANK YOU for your help with this! I knew that I'd have to get later versions of CUDA working eventually, and I wasn't sure how I was going to do it. I am very grateful that you came along and assisted me with this, @jakirkham!

**Comment by jakirkham** [comment]
That's great news! 🎉

Do the environments where this is being installed already have OpenMM 8.0 for some reason? Asking as the example I ran above grabbed OpenMM 8.1 when creating a fresh environment

Happy to help 🙂

<hr>

Also this fixed the bot error referenced in issue ( https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/issues/13 ). So the migrator bot has started adding PRs again

* https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/pull/16
* https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/pull/15

The CUDA 11.8 PR passes and should be mergeable without changes

The CUDA 12.0 PR likely needs changes (is failing currently). Though can follow up on that after CUDA 11.8 is in

Please let me know if you have questions on either of these

**Comment by lvotapka** [comment]
Do the environments where this is being installed already have OpenMM 8.0 for some reason?

No, it did not. I'm not sure why it grabbed that version. That's interesting that it didn't grab 8.0 for yours. Let me do some more testing and figure out what might be causing this.

As for the other PRs, sure let me merge the one for CUDA 11.8, and then I'll see if I can figure out how to fix the issues for CUDA 12. It's possible that OpenMM does not yet work on CUDA 12.

**Comment by jakirkham** [comment]
Interesting, please let me know what you find out

Sounds good

With CUDA 12, there are some more manual changes as more of the CTK is packaged now, but it is also split into smaller components. For general background please see issue ( https://github.com/conda-forge/conda-forge.github.io/issues/1963 ). Can comment on the PR with relevant details

OpenMM has been migrated ( https://github.com/conda-forge/openmm-feedstock/pull/121 ). The migrator bot won't open a PR on a feedstock until all of its dependencies have been migrated

**Comment by jakirkham** [comment]
The migration bot refreshed the CUDA 12 PR (after the CUDA 11.8 PR was merged). Have added some thoughts on next steps in comment: https://github.com/conda-forge/seekr2_openmm_plugin-feedstock/pull/17#issuecomment-1841904745
